tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java

```diff
@@ -0,0 +1,683 @@
+package org.tensorflow.framework.losses;
+
+import org.tensorflow.DataType;
+import org.tensorflow.Operand;
+import org.tensorflow.framework.losses.impl.LossesImpl;
+import org.tensorflow.framework.losses.impl.Tuple;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.core.ReduceAll;
+import org.tensorflow.op.core.ReduceMax;
+import org.tensorflow.op.core.ReduceSum;
+import org.tensorflow.op.core.Variable;
+import org.tensorflow.op.math.Mean;
+import org.tensorflow.op.math.Sigmoid;
+import org.tensorflow.op.math.Softplus;
+import org.tensorflow.op.nn.Softmax;
+import org.tensorflow.types.TBool;
+import org.tensorflow.types.TInt64;
+import org.tensorflow.types.family.TNumber;
+
+/** Built-in loss functions. */
+public class Losses {
+
+  /** Default Fuzz factor. */
+  public static final float EPSILON = 1e-7f;
+
+  /**
+   * Calculates the mean absolute error between labels and predictions.
+   *
+   * <p><code>loss = reduceMean(abs(labels - predictions))</code>
+   *
+   * @param tf The TensorFlow Ops
+   * @param labels the labels
+   * @param predictions the predictions
+   * @param <T> the data type of the result
+   * @param <U> the data type of the labels
+   * @return the mean absolute error
+   */
+  public static <T extends TNumber, U extends TNumber> Operand<T> meanAbsoluteError(
+      Ops tf, Operand<U> labels, Operand<T> predictions) {
+    Operand<T> tLabels = tf.dtypes.cast(labels, predictions.asOutput().dataType());
```

Comment:
Do we want to avoid this cast in the case where `labels` already has the same data type?

Reply:
I guess the question is what is the overhead of casting onto oneself vs the overhead of checking?  I would hope that `tf.dtypes.cast` already handles this, but I could be mistaken.

Reply:
The code for checking could be something like this:
~~~ Java
@SuppressWarnings("unchecked")
private static <T extends TNumber, U extends TNumber> Operand<T> castIfNecessary(
    Operand<U> value, DataType<T> requiredType) {
  return (value.asOutput().dataType() == requiredType) 
      ? (Operand<T>) value
      : tf.dtypes.cast(value, requiredType);
}
~~~
So the overhead of checking would be the function call plus `value.asOutput().dataType() == requiredType`.

Looking at the code for `tf.dtypes.cast`, unless we think a cast is almost always needed, it would be cheaper to do the check to sometimes avoid it.
~~~ Java
  public <U extends TType, T extends TType> Cast<U> cast(Operand<T> x, DataType<U> DstT,
      Cast.Options... options) {
    return Cast.create(scope, x, DstT, options);
  }

  @Endpoint(describeByClass = true)
  public static <U extends TType, T extends TType> Cast<U> create(Scope scope, Operand<T> x, DataType<U> DstT, Options... options) {
    OperationBuilder opBuilder = scope.env().opBuilder("Cast", scope.makeOpName("Cast"));
    opBuilder.addInput(x.asOutput());
    opBuilder = scope.applyControlDependencies(opBuilder);
    opBuilder.setAttr("DstT", DstT);
    if (options != null) {
      for (Options opts : options) {
        if (opts.Truncate != null) {
          opBuilder.setAttr("Truncate", opts.Truncate);
        }
      }
    }
    return new Cast<U>(opBuilder.build());
  }
~~~

Reply:
I like `castIfNecessary` as a general util method. It would be used almost everywhere, so it would be a huge change.
Perhaps create a new PR for `castIfNecessary`, then once that is merged we can start retrofitting all packages under framework.

Reply:
In graph construction mode, an unnecessary call to `cast` creates an unnecessary graph operation.

Reply:
*shrug* it'll be a no-op most of the time and compiled away if we get XLA working. Given the relative size of the computation around it I suspect it won't be an issue.

Reply:
I also vote for a explicit check in the code to avoid adding an extra operation to the graph when it is not required

Reply:
OK, I will add a helper class in` org.tensorflow.framework.utils`, then retrofit the `Loss` classes.

Reply:
Just a comment on @deansher proposed method here, the datatypes for `<U>` and `<T>` should not be restricted to `TNumber` because it is valid to cast to/from `TNumber` and `TBool` .